lib/elixir/lib/access.ex

```diff
@@ -700,4 +700,158 @@ defmodule Access do
   defp get_and_update_at([], _index, _next, updates) do
     {nil, :lists.reverse(updates)}
   end
+
+  @doc ~S"""
+    Returns a function that accesses all elements of a list that match the provided predicate.
+
+    The returned function is typically passed as an accessor to `Kernel.get_in/2`,
+    `Kernel.get_and_update_in/3`, and friends.
+
+    ## Examples
+
+        iex> list = [%{name: "john", salary: 10}, %{name: "mary", salary: 20}, %{name: "francine", salary: 30}]
+        iex> get_in(list, [Access.filter(&(&1.salary >= 20)), :name])
+        ["mary", "francine"]
+        iex> get_and_update_in(list, [Access.filter(&(&1.salary <= 20)), :name], fn
+        ...>   prev -> {prev, String.upcase(prev)}
+        ...> end)
+        {["john", "mary"], [%{name: "JOHN", salary: 10}, %{name: "MARY", salary: 20}, %{name: "francine", salary: 30}]}
+
+    `filter/1` can also be used to pop elements out of a list or
+    a key inside of a list:
+
+        iex> list = [%{name: "john", salary: 10}, %{name: "mary", salary: 20}, %{name: "francine", salary: 30}]
+        iex> pop_in(list, [Access.filter(&(&1.salary >= 20))])
+        {[%{name: "mary", salary: 20}, %{name: "francine", salary: 30}], [%{name: "john", salary: 10}]}
+        iex> pop_in(list, [Access.filter(&(&1.salary >= 20)), :name])
+        {["mary", "francine"], [%{name: "john", salary: 10}, %{salary: 20}, %{salary: 30}]}
+
+    When no match is found, an empty list is returned and the update function is never called
+
+        iex> list = [%{name: "john", salary: 10}, %{name: "mary", salary: 20}, %{name: "francine", salary: 30}]
+        iex> get_in(list, [Access.filter(&(&1.salary >= 50)), :name])
+        []
+        iex> get_and_update_in(list, [Access.filter(&(&1.salary >= 50)), :name], fn
+        ...>   prev -> {prev, String.upcase(prev)}
+        ...> end)
+        {[], [%{name: "john", salary: 10}, %{name: "mary", salary: 20}, %{name: "francine", salary: 30}]}
+
+    An error is raised if the predicate is not a function or is of the incorrect arity:
+
+        iex> get_in([], [Access.filter(fn a, b -> a == b end)])
+        ** (FunctionClauseError) no function clause matching in Access.filter/1
+
+    An error is raised if the accessed structure is not a list:
+
+        iex> get_in(%{}, [Access.filter(fn a -> a == 10 end)])
+        ** (RuntimeError) Access.filter/1 expected a list, got: %{}
+  """
+  def filter(func) when is_function(func, 1) do
```

Comment:
It is better to not check the arity because that leads to a better error later on the invocation site.

Reply:
Good point, I'll change now.